rfc/state-file.md

```diff
@@ -0,0 +1,94 @@
+# Constellation state file
+
+During a cluster's lifecycle, Constellation needs to keep track of multiple different values and state information of the cluster,
+like loadbalancer IPs or UID of the cluster.
+
+If this information is configurable by a user, and required to set up the cluster, the user saves it in Constellation's config file.
+Information that is not configurable, and that is generated by the cluster during the creation or init phase is either saved in Terraform output files,
+or in the cluster ID file (`constellation-id.json`).
+
+This creates problems since the relevant information may be split over multiple files, and it is not always clear at what point certain information is available in the cluster ID file.
+Additionally, commands like `upgrade` and `init` currently depend on Terraform files being present in the working directory,
+which may not be the case if the cloud resources were manually created.
+
+## Proposal
+
+Add a Constellation state file to keep track of this information.
+This file should replace the current `constellation-id.json` file and be the single source of truth for this kind of cluster relevant information.
+The file should keep track of all information currently kept in `constellation-id.json`, as well as information we currently have to read from Terraform files
+during `init` or `upgrade`.
+
+The file should be considered "read-only" by users, if Constellation's cloud resource were created using the Constellation CLI.
+If the cloud resources were created manually by a user, they will need to manually fill in some of the state file's fields.
+
+This will allow us to more easily decouple resource creation/management from Constellation.
+Meaning it should be much easier for Constellation users to have different parties create and update cloud resources and IAM using their own tooling (e.g. Terraform),
+and only use the Constellation CLI for Kubernetes cluster management (Kubernetes versions, service versions, image upgrades etc.).
+
+## Format
+
+The state file will be in either YAML or JSON format, and contain a version number to keep track of necessary migrations or breaking formats.
+The file should be split into multiple sections for each relevant resource.
+
+Information that is only used to upgrade Terraform resources (e.g. IAM) should not be part of this file.
+
+The following is a list of proposed sections:
```

Comment:
Rename to "Writeable" and "Readonly"? It's a bit clumsy, but should make things obvious. And I feel like the current categories aren't that precisely defined, yet.

Reply:
Writeable and Readonly are IMO not a good fit, since this would only describe how a user can interact with them, but not what these resources actually are.
The name of these keys should describe accurately what these resources are and ideally also at what point in Constellation's lifecycle they are available.
Part of the problem with the current id file is that we write multiple different values to the same struct, but at different times.
This makes it unclear during which command what information is expected to be available from the id file.
The state file should fix this issue, and I dont think Writeable and Readonly would do that.